### PR TITLE
Implement backward-compatible closure capture behavior with parser lookup disabled

### PR DESCRIPTION
**Pull request by slavapestov**
I wrote about the existing Swift scoping rules for local functions here: https://forums.swift.org/t/clarify-scoping-behavior-with-local-functions/40558

The quick summary is that today, parser lookup runs first, and only finds names that precede the use location in the source file. The old context-based lookup done from `preCheckExpression()` finds all local declarations, even those not in scope. A set of heuristics are used to find the relevant local declaration.

Since ASTScope has to replace parse-time lookup, it has to model both behaviors. This PR wires it all up.

In the future, we can consider simplifying pre-check's name lookup rules by banning forward references to captured bindings, as I wrote in my forum post. But that's not in the critical path for removing parser lookup anymore.

Builds upon https://github.com/apple/swift/pull/34125.

### CODE REVIEW

**Comment by slavapestov** [comment]
Note: The large diff is primarily because of the commit that moves preCheckExpression() into it's own file. There are no other changes in that specific commit, so you can review the others individually instead.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/AST/ASTScopeCreation.cpp [hunk 1]
```diff
@@ -667,10 +667,30 @@ class NodeAdder
...
+        if (auto *vd = dyn_cast<ValueDecl>(localBinding)) {
+          if (isa<FuncDecl>(vd)  || isa<TypeDecl>(vd)) {
+            localFuncsAndTypes.push_back(vd);
+          } else if (auto *var = dyn_cast<VarDecl>(localBinding)) {
+            localVars.push_back(var);
+          }
+        }
+      }
+    }
+
```
Comment (davidungar) [inline lib/AST/ASTScopeCreation.cpp:687]: Nice and clear!

File: lib/AST/ASTScopeLookup.cpp [hunk 2]
```diff
@@ -371,20 +368,16 @@ bool DifferentiableAttributeScope::lookupLocalsOrMembers(
...
+  if (consumer.consume(localFuncsAndTypes))
+    return true;
+
+  if (consumer.consumePossiblyNotInScope(localVars))
+    return true;
+
+  if (consumer.finishLookupInBraceStmt(stmt))
+    return true;
+
+  return false;
```
Comment (davidungar) [inline lib/AST/ASTScopeLookup.cpp:380]: Very interesting! Looks good here.

**Review by davidungar — APPROVED** [review]
I focused on the name lookup part, and that looks very good! As you warned, there is a huge, artificial diff in PreCheckExpr. I'm not sure what if anything changed there. Great to be cleaning things up this way!

**Comment by slavapestov** [comment]
@davidungar I split out that change in https://github.com/apple/swift/pull/34152

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by davidungar** [comment]
Still looks fine!

**Comment by compnerd** [comment]
@slavapestov I think that this change is causing some issues with the latest changes that @hborla did: https://ci-external.swift.org/job/oss-swift-windows-x86_64-vs2019/3047/console shows a couple of test failures (that I also saw on Linux - I don't think that this is Windows specific).

**Comment by slavapestov** [comment]
Oops, my bad. Thanks for reverting. I'll take a look.

File: lib/AST/UnqualifiedLookup.cpp [hunk 3]
```diff
@@ -587,6 +604,21 @@ bool ASTScopeDeclConsumerForUnqualifiedLookup::consume(
...
+  for (auto *var : vars) {
+    if (!factory.Name.getFullName().isSimpleName(var->getName()))
+      continue;
+
+    factory.Results.push_back(LookupResultEntry(var));
+  }
+
+  return false;
+}
+
```
Comment (AnthonyLatsis) [inline lib/AST/UnqualifiedLookup.cpp:609]: Was it intentional to always return false here?

**Pull request merged**